store/src/java/com/zimbra/cs/mailbox/Mailbox.java

```diff
@@ -1105,74 +1105,79 @@ public void addListener(Session session) throws ServiceException {
             return;
         }
         assert(session.getSessionId() != null);
+        if (maintenance != null) {
+            throw MailServiceException.MAINTENANCE(mId);
+        }
+        ZimbraLog.mailbox.debug("adding listener: %s", session);
+        if (!mListeners.contains(session)) {
+            mListeners.add(session);
+        }
+        /* check whether beginMaintenance happened whilst adding a listener.  If it did then
+         * undo it.  This avoids getting a write lock */
+        if (maintenance != null) {
```

Comment:
Could we combine this maintenance check with the previous one?
Or perhaps remove the first one and let them all be handled here?

Reply:
writes to mListener involve discarding the previous array and creating a new one, so avoiding this is worth doing - so I'd rather not just rely on the 2nd if.  With 2 ifs, the 2nd check should  hardly ever succeed (but it will sometimes)  I consider it is better to have this 2 stage check than to lock the mailbox.

Reply:
Fair enough. It was just a thought.